test/validation/api/ipsec/ipsec.c

```diff
@@ -1153,6 +1200,23 @@ int ipsec_config(odp_instance_t ODP_UNUSED inst)
 	ipsec_config.inbound.chksums.all_chksum = ~0;
 	ipsec_config.stats_en = true;
 
+	ipsec_config.inbound.reassembly.max_wait_time = 20 * ODP_TIME_MSEC_IN_NS;
+	ipsec_config.inbound.reassembly.max_num_frags = capa.reassembly.max_num_frags;
+
+	if (suite_context.reass_ipv4 && (capa.reassembly.ip || capa.reassembly.ipv4))
+		ipsec_config.inbound.reassembly.en_ipv4 = true;
```

Comment:
This seems to be wrong. If capa.reassembly.ip is set but capa.reassembly.ipv4 is not set, then reassembly of only IPv4 but not IPv6 is not supported. But this code configures reassembly for IPv4 only.

Maybe this error is an indication that the capability bits are not clear after all and should be reworked.

Reply:
```
	/** Reassembly offload for both IPv4 and IPv6 packets */
	odp_bool_t ip;
This conversation was marked as resolved by anoobj

	/** Reassembly offload for IPv4 packets only */
	odp_bool_t ipv4;

	/** Reassembly offload for IPv6 packets only */
	odp_bool_t ipv6;
```
If capa.reassembly.ip is set, then both ipv4 & ipv6 is supported. capa.reassembly.ipv4 & capa.reassembly.ipv6 is applicable only when capa.reassembly.ip is not set.

Reply:
But only at the same time.

My understanding of the API proposal is this:

```
        /** Reassembly offload for both IPv4 and IPv6 packets */
        odp_bool_t ip;
```
This means that reassemly offload can be enabled for both IPv4 and IPv6 packets at the same time. This does not tell whether the offload can be enabled for only one or the other.

```
        /** Reassembly offload for IPv4 packets only */
        odp_bool_t ipv4;
```
This means that reassembly offload can be enabled for IPv4 only, leaving it disabled for IPv6.

The code I commented may enable reassembly offload for IPv4 only, when the offload is supported only for IPv4 and IPv6 both at the same time.

Regardless of what the right interpretation is now, the API needs to be made clearer so that potential users cannot easily understand it in a different way than intended.

Reply:
@JannePeltonen Let me summarize what I understood. The capabilities say what is supported by the implementation. It can be
1. Both IPv4 & IPv6
2. IPv4 only
3. IPv6 only

Now, in the config, we have two options,
1. Enable for IPv4
2. Enable for IPv6

If caps say, both IPv4 & IPv6, then you suggest enabling both IPv4 & IPv6? Irrespective of the flags provided by the caller? I don't think we need a rewording in spec for this. I can make this adjustment in UT.

Reply:
@JannePeltonen Or if you think rewording is required, please suggest inline in spec patch.

Reply:
> If caps say, both IPv4 & IPv6, then you suggest enabling both IPv4 & IPv6? Irrespective of the flags provided by the caller?

Yes, if "IPv4 only" is not supported.

If caps say that "IPv4 only" is not supported, then the test application should not try to configure IPv4 only. I'll see if I can come up with a rewording proposal for the API to clarify it.

Reply:
> If caps say that "IPv4 only" is not supported, then the test application should not try to configure IPv4 only. I'll see if I can come up with a rewording proposal for the API to clarify it.

How about this:

```
 typedef struct odp_reass_capability_t {
-       /** Reassembly offload for both IPv4 and IPv6 packets */
+       /** Reassembly offload for both IPv4 and IPv6 packets. This capability
+        *  does not allow enabling reassembly for only IPv4 or only IPv6.
+        */
        odp_bool_t ip;

-       /** Reassembly offload for IPv4 packets only */
+       /** Reassembly offload for IPv4 packets */
        odp_bool_t ipv4;

-       /** Reassembly offload for IPv6 packets only */
+       /** Reassembly offload for IPv6 packets */
        odp_bool_t ipv6;

        /** Maximum time in ns that a fragment can wait in the reassembly
@@ -86,10 +88,18 @@ typedef struct odp_reass_capability_t {
  * @see odp_packet_reass_status(), odp_packet_reass_partial_state()
  */
 typedef struct odp_reass_config_t {
-       /** Attempt inline reassembly of IPv4 packets. Disabled by default */
+       /** Attempt inline reassembly of IPv4 packets. Disabled by default.
+        *  This may be set if the relevant odp_reass_capability_t::ipv4
+        *  capability is present or if odp_reass_capability_t::ip capability
+        *  is present and en_ipv6 is also set.
+        */
        odp_bool_t en_ipv4;

-       /** Attempt inline reassembly of IPv6 packets. Disabled by default */
+       /** Attempt inline reassembly of IPv6 packets. Disabled by default.
+        *  This may be set if the relevant odp_reass_capability_t::ipv6
+        *  capability is present or if odp_reass_capability_t::ip capability
+        *  is present and en_ipv4 is also set.
+        */
        odp_bool_t en_ipv6;

```

Reply:
Will address as part of v18.